### PR TITLE
Tests: cardano slip39 device tests

### PR DESCRIPTION
**Pull request by ciny**
This PR implements #367 
the test vectors for sign_tx were generated using [this code](https://repl.it/@stanociny/JuicyShamefulAbstractions), you can uncomment various parts to verify the vectors we use. 

note: INVALID_VECTORS test in sign_tx worked without changes (or using a passphrase with slip39). Maybe it would be worth looking into whether we're testing what we think we're testing.

### CODE REVIEW

**Comment by matejcik** [comment]
> note: INVALID_VECTORS test worked without changes (or using a passphrase with slip39). Maybe it would be worth looking into whether we're testing what we think we're testing.

what changes were required for valid vector tests? signature changes?
if so, ISTM the invalid vector tests are right to continue working, as the signature (which would be the difference) is not produced.

Speaking of which, do we actually need to test signature generation? Wouldn't addresses be enough (to prove that the correct key pairs are generated)?

**Comment by ciny** [comment]
> what changes were required for valid vector tests? signature changes?

and "tx_hash" (id) in case of change in outputs. But you are right, looking more closely we're expecting output address validation to fail which isn't affected by the loaded seed. In that case we probably don't need those tests when testing shamir. 

> Wouldn't addresses be enough (to prove that the correct key pairs are generated)?

That's probably a question for @andrewkozlik

**Comment by andrewkozlik** [comment]
> > what changes were required for valid vector tests? signature changes?
> 
> and "tx_hash" (id) in case of change in outputs. But you are right, looking more closely we're expecting output address validation to fail which isn't affected by the loaded seed. In that case we probably don't need those tests when testing shamir.

Agreed, we don't need invalid vector tests for Shamir.

> > Wouldn't addresses be enough (to prove that the correct key pairs are generated)?

I am not sure whether checking the address is entirely sufficient. I doubt that the address depends on the private key extension, but I didn't look into this. If it doesn't, then the tests wouldn't be able to detect if we got the extension wrong. Checking the extended private key would probably be sufficient, but I assume that that shouldn't be possible in device tests. So I think the most reliable solution is to check the signature itself.

**Comment by ciny** [comment]
> > > what changes were required for valid vector tests? signature changes?
> > 
> > 
> > and "tx_hash" (id) in case of change in outputs. But you are right, looking more closely we're expecting output address validation to fail which isn't affected by the loaded seed. In that case we probably don't need those tests when testing shamir.
> 
> Agreed, we don't need invalid vector tests for Shamir.

I removed them.

**Comment by matejcik** [comment]
> > > Wouldn't addresses be enough (to prove that the correct key pairs are generated)?
> 
> I am not sure whether checking the address is entirely sufficient. I doubt that the address depends on the private key extension, but I didn't look into this. If it doesn't, then the tests wouldn't be able to detect if we got the extension wrong. Checking the extended private key would probably be sufficient, but I assume that that shouldn't be possible in device tests. So I think the most reliable solution is to check the signature itself.

how about checking _public_ keys then? cardano exposes both public key and chain code.

What bothers me about the signing test is that it doesn't seem to test anything that is Shamir-specific. If this test fails, either non-Shamir signing test is also failing, or the Shamir key derivation tests are also failing.
(unless there are slip23 modifications in the *signing* code?)

**Comment by andrewkozlik** [comment]
> > > > Wouldn't addresses be enough (to prove that the correct key pairs are generated)?
> > 
> > 
> > I am not sure whether checking the address is entirely sufficient. I doubt that the address depends on the private key extension, but I didn't look into this. If it doesn't, then the tests wouldn't be able to detect if we got the extension wrong. Checking the extended private key would probably be sufficient, but I assume that that shouldn't be possible in device tests. So I think the most reliable solution is to check the signature itself.
 
> how about checking _public_ keys then? cardano exposes both public key and chain code.

Neither the public key nor the chain code depend on the private key extension, so we wouldn't be able to tell if we got it wrong.

> What bothers me about the signing test is that it doesn't seem to test anything that is Shamir-specific. If this test fails, either non-Shamir signing test is also failing, or the Shamir key derivation tests are also failing.

The signing test checks that the SLIP-0023 code derived the private key extension correctly. This wouldn't be detected in any non-Shamir signing test. The Shamir key derivation device tests also wouldn't detect it, because we don't check (and AFAIK can't check) the private key extension in device tests.

> (unless there are slip23 modifications in the _signing_ code?)

SLIP-0023 doesn't modify the signing operations.

Note that by the term *private key extension* I am referring to *k*<sub>R</sub> in https://cardanolaunch.com/assets/Ed25519_BIP.pdf and https://github.com/satoshilabs/slips/blob/master/slip-0023.md.

File: python/trezorlib/tests/device_tests/test_msg_cardano_get_address_shamir.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        (
+            "m/44'/1815'/0'/0/2",
+            "Ae2tdPwUPEZGXmSbda1kBNfyhRQGRcQxJFdk7mhWZXAGnapyejv2b2U3aRb",
+        ),
+    ],
+)
+@setup_client(mnemonic=SLIP39_MNEMONIC, passphrase=True)
+def test_cardano_get_address(client, path, expected_address):
+    # enter passphrase
+    assert client.debug.read_passphrase_protection() is True
```
Comment (matejcik) [inline python/trezorlib/tests/device_tests/test_msg_cardano_get_address_shamir.py]: is there a reason to do this, as opposed to `assert client.features.passphrase_protection is True` ?

Comment (ciny) [reply python/trezorlib/tests/device_tests/test_msg_cardano_get_address_shamir.py]: not that I can think of, will fix.

**Pull request merged**